File: python-sdk/src/dispersl/ndjson.py
from __future__ import annotations

import json
from collections.abc import AsyncIterator

from .errors import StreamParseError
from .models import NDJSONChunk


async def parse_ndjson_stream(lines: AsyncIterator[str]) -> AsyncIterator[NDJSONChunk]:
    buffer = ""
    async for part in lines:
        buffer += part
        split_lines = buffer.split("\n")
        buffer = split_lines.pop() if split_lines else ""
        for raw in split_lines:
            raw = raw.strip()
            if not raw:
                continue
            try:
                yield NDJSONChunk.model_validate(json.loads(raw))
            except Exception as exc:
                raise StreamParseError(f"Failed to parse NDJSON line: {raw}") from exc

    tail = buffer.strip()
    if tail:
        try:
            yield NDJSONChunk.model_validate(json.loads(tail))
        except Exception as exc:
            raise StreamParseError(f"Failed to parse NDJSON tail: {tail}") from exc

File: typescript-sdk/src/ndjson.ts
import type { NDJSONChunk } from "./types";
import { StreamParseError } from "./errors";

export class NdjsonParser {
  async parse(
    stream: ReadableStream<Uint8Array>,
    onChunk?: (chunk: NDJSONChunk) => void | Promise<void>
  ): Promise<void> {
    const reader = stream.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() ?? "";

        for (const line of lines) {
          if (!line.trim()) continue;
          try {
            const parsed = JSON.parse(line) as NDJSONChunk;
            if (onChunk) await onChunk(parsed);
          } catch (error) {
            throw new StreamParseError(`Failed parsing NDJSON line: ${line}`, String(error));
          }
        }
      }

      if (buffer.trim()) {
        try {
          const parsed = JSON.parse(buffer) as NDJSONChunk;
          if (onChunk) await onChunk(parsed);
        } catch (error) {
          throw new StreamParseError(`Failed parsing NDJSON tail: ${buffer}`, String(error));
        }
      }
    } finally {
      reader.releaseLock();
    }
  }
}
